File: Multi-threaded Data Processing System/MultiThreadedSystem.java
import java.io.*;
import java.util.*;
import java.util.concurrent.*;
public class RideSharingSystem {
    private static final int NUM_WORKERS = 4;
    private static final int NUM_TASKS = 10;
    private static final String OUTPUT_FILE = "output_java.txt";

    public static void main(String[] args) throws InterruptedException, IOException {
        BlockingQueue<Task> taskQueue = new LinkedBlockingQueue<>();
        List<String> results = Collections.synchronizedList(new ArrayList<>());

        // Add tasks to the queue
        for (int i = 1; i <= NUM_TASKS; i++) {
            taskQueue.add(new Task(i));
        }

        ExecutorService executor = Executors.newFixedThreadPool(NUM_WORKERS);
        for (int i = 0; i < NUM_WORKERS; i++) {
            executor.submit(new Worker(taskQueue, results));
        }

        executor.shutdown();
        executor.awaitTermination(1, TimeUnit.MINUTES);

        // Write results to output file
        try (BufferedWriter writer = new BufferedWriter(new FileWriter(OUTPUT_FILE))) {
            for (String result : results) {
                writer.write(result);
                writer.newLine();
            }
        } catch (IOException e) {
            System.err.println("Error writing to file: " + e.getMessage());
        }
    }

    static class Task {
        private final int id;

        public Task(int id) {
            this.id = id;
        }

        public int getId() {
            return id;
        }
    }

    static class Worker implements Runnable {
        private final BlockingQueue<Task> taskQueue;
        private final List<String> results;

        public Worker(BlockingQueue<Task> queue, List<String> results) {
            this.taskQueue = queue;
            this.results = results;
        }

        @Override
        public void run() {
            while (true) {
                try {
                    Task task = taskQueue.poll(1, TimeUnit.SECONDS);
                    if (task == null) break;

                    // Simulate work
                    Thread.sleep(1000);
                    String result = "Processed task: " + task.getId() + " by " + Thread.currentThread().getName();
                    results.add(result);

                    System.out.println("INFO: " + result);
                } catch (InterruptedException e) {
                    Thread.currentThread().interrupt();
                    System.err.println("Thread interrupted: " + e.getMessage());
                    break;
                } catch (Exception e) {
                    System.err.println("Exception occurred: " + e.getMessage());
                }
            }
        }
    }
}

File: Multi-threaded Data Processing System/MultiThreadedSystem.go
// Go Implementation of Ride Sharing Data Processing System
package main

import (
	"fmt"
	"log"
	"os"
	"sync"
	"time"
)

type Task struct {
	ID int
}

func worker(id int, tasks <-chan Task, results *[]string, mu *sync.Mutex, wg *sync.WaitGroup) {
	defer wg.Done()
	for task := range tasks {
		time.Sleep(1 * time.Second) // Simulate processing time
		result := fmt.Sprintf("Processed task: %d by worker-%d", task.ID, id)
		mu.Lock()
		*results = append(*results, result)
		mu.Unlock()
		log.Println("INFO:", result)
	}
}

func main() {
	numWorkers := 4
	numTasks := 10
	taskChan := make(chan Task, numTasks)
	var results []string
	var mu sync.Mutex
	var wg sync.WaitGroup

	// Start workers
	for i := 1; i <= numWorkers; i++ {
		wg.Add(1)
		go worker(i, taskChan, &results, &mu, &wg)
	}

	// Add tasks to the channel
	for i := 1; i <= numTasks; i++ {
		taskChan <- Task{ID: i}
	}
	close(taskChan)

	wg.Wait()

	// Write results to file
	file, err := os.Create("output_go.txt")
	if err != nil {
		log.Fatalf("Error creating file: %v", err)
	}
	defer file.Close()

	for _, result := range results {
		_, err := fmt.Fprintln(file, result)
		if err != nil {
			log.Printf("Error writing to file: %v", err)
		}
	}
}
